### PR TITLE
No sensor#1074

### PR DESCRIPTION
**Pull request by dennisguse**
I missed a lot of stuff for #500.
This is the rest.

Test coverage is still low.

### CODE REVIEW

File: src/main/java/de/dennisguse/opentracks/stats/TrackStatisticsUpdater.java [hunk 1]
```diff
@@ -148,23 +148,17 @@ public void addTrackPoint(TrackPoint trackPoint, Distance minGPSDistance) {
         if (trackPoint.hasSensorDistance()) {
             // Sensor-based distance/speed
             currentSegment.addTotalDistance(trackPoint.getSensorDistance());
-        } else if (lastTrackPoint != null && trackPoint.isMoving()) {
+        } else if (lastTrackPoint != null && lastTrackPoint.hasLocation() && trackPoint.isMoving()) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/stats/TrackStatisticsUpdater.java:151]: We might still have some issue here with mixing SENSORPOINT and TRACKPOINT as distanceTo does not work if between GPS-only TRACKPOINT and SENSORPOINT.
Just a gut feeling.

Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/stats/TrackStatisticsUpdater.java:151]: Or do we just ignore the distance in this case?

Comment (rgmf) [reply src/main/java/de/dennisguse/opentracks/stats/TrackStatisticsUpdater.java:151]: BTW (Offtopic): `minGPSDistance` argument in this method is unused.

**Comment by dennisguse** [comment]
@rgmf @pstorch ready for testing aka it did not crash on  a the 3rd test drive and distance looked okay ;D

**Comment by rgmf** [comment]
> @rgmf @pstorch ready for testing aka it did not crash on a the 3rd test drive and distance looked okay ;D

Installing this branch... and ready for testing ;)

**Comment by rgmf** [comment]
@dennisguse in a 15km route by car it worked like a charm, comparing the result with fdroid version.

I didn't use any sensor in this test.

**Comment by dennisguse** [comment]
@rgmf thanks!
Will check with sensors as well :)

PS/ sensor that provide distance are now special as they should lead to creation of SENSORPOINTs while heart rate would not.

**Comment by pstorch** [comment]
I've also installed this branch. Let's see, if I have the chance to test it in the next few days.

**Comment by rgmf** [comment]
@dennisguse running test with HR sensor: it worked like a charm :partying_face:

**Comment by rgmf** [comment]
@dennisguse @pstorch I keep testing it. This time with hr, cadence and speed/distance sensors. Bad news: after 6km (by bike) OT crashed with this error:

Reading the logs I don't know if it's related with this PR or a general issue.

```log
1-06 08:49:45.003 20085 20085 E AndroidRuntime: FATAL EXCEPTION: main
01-06 08:49:45.003 20085 20085 E AndroidRuntime: Process: de.dennisguse.opentracks.debug, PID: 20085
01-06 08:49:45.003 20085 20085 E AndroidRuntime: java.lang.RuntimeException: Moving time cannot be negative
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.stats.TrackStatistics.addMovingTime(TrackStatistics.java:211)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.stats.TrackStatistics.addMovingTime(TrackStatistics.java:205)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.stats.TrackStatisticsUpdater.addTrackPoint(TrackStatisticsUpdater.java:163)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.TrackRecordingManager.insertTrackPointHelper(TrackRecordingManager.java:229)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.TrackRecordingManager.insertTrackPoint(TrackRecordingManager.java:223)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.TrackRecordingManager.onNewTrackPoint(TrackRecordingManager.java:189)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.TrackRecordingService.newTrackPoint(TrackRecordingService.java:373)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.handlers.TrackPointCreator.onNewTrackPoint(TrackPointCreator.java:125)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at de.dennisguse.opentracks.services.handlers.GPSHandler.onLocationChanged(GPSHandler.java:124)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.location.LocationListener.onLocationChanged(LocationListener.java:63)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.location.LocationManager$LocationListenerTransport$1.operate(LocationManager.java:3180)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.location.LocationManager$LocationListenerTransport$1.operate(LocationManager.java:3177)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at com.android.internal.listeners.ListenerExecutor.lambda$executeSafely$0(ListenerExecutor.java:127)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at com.android.internal.listeners.ListenerExecutor$$ExternalSyntheticLambda0.run(Unknown Source:8)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7839)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
01-06 08:49:45.003 20085 20085 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
```

**Comment by dennisguse** [comment]
@rgmf Sadly, this problem already exists in v3.23.0 - I get error reports via Google Play.
Reason: before that negative moving time was just ignored, and I moved this to a runtime exception.
Ugly thing: it seems to be a race condition OR the system clock is changed (i.e., NTP network synchronization).
My current guess: we replace the timestamps for GPS provided data.
* race condition: UI timer and GPS creation interfere
* clock update see #800
 
If it is clock update, then I made this issue just occur more often and it was happening before already, but we just ignored it.
So far this only happened when the clock was set to a time before currentSegment.startTime..

**Comment by pstorch** [comment]
I took it to a test ride today. By bike with two intermediate stops. At the stops I unlocked my phone and saw the Screen shortly with some data and then resetting. Looks like a crash and reopening. It only recorded a few meters, but I traveled some km. 
Unfortunately on the road I don't have access to the logcat. So I don't know if it is the same crash you already discussed.

**Comment by dennisguse** [comment]
@rgmf @pstorch My proposal: we revert the RuntimeException back to ignore negative moving time, merge this, and check if it still crashes.
If it doesn't we release v3.24.0  - as we have some issues in production I would like to get rid of them soonish.
And then I will take care about #800 and implement a monotonic clock.
How does this sound?

**Comment by dennisguse** [comment]
Today's nightly worked me; crafting the release :D

**Comment by pstorch** [comment]
Used the nightly today as well for my workout (despite your warning). No crash 😀

**Comment by dennisguse** [comment]
After merging this PR, it became okayish. Still the distance sensors might give us some trouble, but that's how it goes :D

**Pull request merged**